### PR TITLE
Clean up unnecessary portions of Foundation overlay

### PR DESCRIPTION
**Pull request by itaiferber**
**What's in this pull request?**
Prunes some now-unnecessary portions of the Foundation overlay. Most of these changes are not strictly needed for ABI stability, but are the result of reviewing the overlay for stability. (rdar://problem/33678210)

### CODE REVIEW

File: stdlib/public/SDK/Foundation/Boxing.swift [hunk 1]
```diff
@@ -62,9 +62,3 @@ extension _MutableBoxing {
         return whatToDo(_handle._pointer)
     }
 }
-
-internal enum _MutableUnmanagedWrapper<ImmutableType : NSObject, MutableType : NSObject>
```
Comment (itaiferber) [inline stdlib/public/SDK/Foundation/Boxing.swift]: This is no longer used — being `internal`, we can just delete it.

File: stdlib/public/SDK/Foundation/Decimal.swift [hunk 2]
```diff
@@ -13,6 +13,8 @@
 @_exported import Foundation // Clang module
 import _SwiftCoreFoundationOverlayShims
 
+// The methods in this extension exist to match the protocol requirements of FloatinPoint, even if we can't conform directly.
+// If it becomes clear that conformance is truly impossible, we can deprecate some of the methods (e.g. isEqual(to:)) in favor of operators.
```
Comment (itaiferber) [inline stdlib/public/SDK/Foundation/Decimal.swift]: Added some comments to highlight why we have these methods. I was looking at `Decimal` to figure out why we need both `isEqual(to:)` and `==` (and `isLess(than:)` and `<`), and the answer is our pseudo-conformance to `FloatingPoint`; both of the methods come from there.

File: stdlib/public/SDK/Foundation/NSCoder.swift [hunk 3]
```diff
@@ -229,16 +229,3 @@ extension NSKeyedUnarchiver {
     }
   }
 }
-
-
-@available(*, deprecated, renamed:"NSCoding", message: "Please use NSCoding")
-typealias Coding = NSCoding
```
Comment (itaiferber) [inline stdlib/public/SDK/Foundation/NSCoder.swift]: These aren't public, so we can just delete them. Oops.
(Confirmed that attempting to use these gives `use of undeclared identifier`)

File: stdlib/public/SDK/Foundation/NSDictionary.swift [hunk 4]
```diff
@@ -34,7 +34,7 @@ extension Dictionary {
   ///
   /// The provided `NSDictionary` will be copied to ensure that the copy can
   /// not be mutated by other code.
-  public init(_cocoaDictionary: __shared _NSDictionary) {
+  fileprivate init(_cocoaDictionary: __shared _NSDictionary) {
```
Comment (itaiferber) [inline stdlib/public/SDK/Foundation/NSDictionary.swift]: This *used* to be an entry point to bridging as far as Philippe and I can remember. This is no longer used anywhere except in this file. (Same with the `NSSet` change.)

**Comment by jrose-apple** [comment]
I'll take a look soon but meanwhile I want to see how much this improves code size. :-)

@swift-ci Please smoke benchmark

File: stdlib/public/SDK/Foundation/NSGeometry.swift [hunk 5]
```diff
@@ -20,35 +20,15 @@ import CoreGraphics
 // NSRectEdge
 //===----------------------------------------------------------------------===//
 
-// In the SDK, the following NS*Edge constants are defined as macros for the
-// corresponding CGRectEdge enumerators.  Thus, in the SDK, NS*Edge constants
-// have CGRectEdge type.  This is not correct for Swift (as there is no
-// implicit conversion to NSRectEdge).
-
-@available(*, unavailable, renamed: "NSRectEdge.MinX")
-public var NSMinXEdge: NSRectEdge {
```
Comment (itaiferber) [inline stdlib/public/SDK/Foundation/NSGeometry.swift]: Foundation's `NSGeometry.h` already has these unavailable messages on the relevant types directly for all platforms except for 32-bit Mac. (There we define the `NS` types in terms of `CG` types, except that 32-bit Mac gets its own definitions.)

File: stdlib/public/SDK/Foundation/NSIndexSet.swift [hunk 6]
```diff
@@ -12,7 +12,7 @@
 
 @_exported import Foundation // Clang module
 
-// FIXME: move inside NSIndexSet when the compiler supports this.
+// TODO: Evaluate deprecating with a message in favor of IndexSet.
```
Comment (itaiferber) [inline stdlib/public/SDK/Foundation/NSIndexSet.swift]: @phausler and I couldn't see much of a benefit to moving this into `NSIndexSet` itself — it's unlikely anyone is referring to this type by name, but if they are, the risk of breakage is kind of silly. In reality, you should probably be using `IndexSet` anyway.

File: stdlib/public/SDK/Foundation/NSObject.swift [hunk 7]
```diff
@@ -13,7 +13,10 @@
 @_exported import Foundation // Clang module
 import ObjectiveC
 
+// This exists to allow for dynamic dispatch on KVO methods added to NSObject.
+// Extending NSObject with these methods would disallow overrides.
```
Comment (itaiferber) [inline stdlib/public/SDK/Foundation/NSObject.swift]: Annotating to save someone the time of figuring out the purpose of this protocol in the future.

File: stdlib/public/SDK/Foundation/NSSet.swift [hunk 8]
```diff
@@ -17,7 +17,7 @@ extension Set {
   ///
   /// The provided `NSSet` will be copied to ensure that the copy can
   /// not be mutated by other code.
-  public init(_cocoaSet: __shared _NSSet) {
+  fileprivate init(_cocoaSet: __shared _NSSet) {
```
Comment (itaiferber) [inline stdlib/public/SDK/Foundation/NSSet.swift]: See `NSDictionary` above.

File: stdlib/public/SDK/Foundation/TypePreservingNSNumber.mm [hunk 9]
```diff
@@ -1,361 +0,0 @@
-//===----------------------------------------------------------------------===//
```
Comment (itaiferber) [inline stdlib/public/SDK/Foundation/TypePreservingNSNumber.mm]: This file was no longer being compiled.

**Comment by itaiferber** [comment]
@swift-ci Please test macOS

**Comment by itaiferber** [comment]
@swift-ci Please smoke test Linux

**Comment by itaiferber** [comment]
@jrose-apple Sounds good. This is not yet a full audit for inlinability though, so we'll likely see some changes there.

File: stdlib/public/SDK/Foundation/Decimal.swift [hunk 10]
```diff
@@ -13,6 +13,8 @@
 @_exported import Foundation // Clang module
 import _SwiftCoreFoundationOverlayShims
 
+// The methods in this extension exist to match the protocol requirements of FloatinPoint, even if we can't conform directly.
```
Comment (jrose-apple) [inline stdlib/public/SDK/Foundation/Decimal.swift]: Typo: "FloatinPoint"

**Review by jrose-apple — APPROVED** [review]
All of this looks good to me; very straightforward.

File: stdlib/public/SDK/Foundation/Decimal.swift (hunk 10 above)
Comment (itaiferber) [reply stdlib/public/SDK/Foundation/Decimal.swift]: Alternatively: we propose to amend IEEE-754 for Floatin' Point numbers

Comment (itaiferber) [reply stdlib/public/SDK/Foundation/Decimal.swift]: (I'll wait for the full macOS test to pass before amending so we can make sure it's green; I'll smoke test afterwards)

**Comment by itaiferber** [comment]
@swift-ci Please test and merge

**Pull request merged**